rfcs/primitive_lists_api.md

```diff
@@ -0,0 +1,283 @@
+public class PrimitiveIntegerRealmQuery extends PrimitiveRealmQuery<Integer>
+public class PrimitiveFloatRealmQuery extends PrimitiveRealmQuery<Float>
+public class PrimitiveDoubleRealmQuery extends PrimitiveRealmQuery<Double>
+public class PrimitiveStringRealmQuery extends PrimitiveRealmQuery<String>
+public class PrimitiveBooleanRealmQuery extends PrimitiveRealmQuery<Boolean>
+
+protected abstract class PrimitiveRealmList<Object> extends OrderedRealmCollection<Object>
+public class PrimitiveIntegerRealmList extends PrimitiveRealmList<Integer>
+public class PrimitiveFloatRealmList extends PrimitiveRealmList<Float>
+public class PrimitiveDoubleRealmList extends PrimitiveRealmList<Double>
+public class PrimitiveStringRealmList extends PrimitiveRealmList<String>
+public class PrimitiveBooleanRealmList extends PrimitiveRealmList<Boolean>
+
+protected abstract class PrimitiveRealmResults<Object> extends OrderedRealmCollection<Object>
+public class PrimitiveIntegerRealmResults extends PrimitiveRealmResults<Integer>
+public class PrimitiveFloatRealmResults extends PrimitiveRealmResults<Float>
+public class PrimitiveDoubleRealmResults extends PrimitiveRealmResults<Double>
+public class PrimitiveStringRealmResults extends PrimitiveRealmResults<String>
+public class PrimitiveBooleanRealmResults extends PrimitiveRealmResults<Boolean>
+```
+
+2) We need the same optional methods in `RealmCollection` and `OrderedRealmCollection` unless we we create an entirely
+   new collection hierarchy.
+
+3) `PrimitiveRealmQuery` will only contain the features supported by primitive arrays. If using 1b, we can even ensure
+   that all methods are useful
+
+```
+beginsWith // Only String
+between // Only numbers, date
+contains // Only String
+distinct
+endsWith // Only String
+equalTo
+findAllSorted
+findAllSortedAsync
+greaterThan // Only numbers, date
+greaterThanOrEqualTo // only numbers, date
+in
+isEmpty
+isNotEmpty
+isNotNull
+isNull
+lessThan // only numbers, date
+lessThanOrEqualTo // only numbers, date
+like // only String
+max // only numbers, date
+maximumDate // only numbers, date
+min // only numbers, date
+minimumDate // only numbers, date
+notEqualTo
+sum // only numbers // only numbers
+````
+
+# Discussion
+
+## Design 1: Relax type requirements
+
+**Advantages**
+* No need to add new types.
+* RealmQuery/RealmResults already contain methods that doesn't work in certain circumstances.
+* Schemas can already be restrict using modules, making `<E extends RealmModel>` inaccurate anyway.
+* RealmCollection will have to contain optional methods no matter what, so adding them to RealmQuery doesn't break
+  that pattern.
+* Would allow us to remove the requirement that people implements `RealmModel`, they could just use the `@RealmClass`
+  annotation.
+
+**Disadvantages**
+* API gets a lot less type safe (IntelliJ plugin might help here).
+* `RealmQuery` will suddenly have a ton of new methods that will only work for primitive arrays (lint check?)
```

Comment:
Right.  I can see that I am in a minority, here, but this is what types are *for*.  I think this argument, all by itself, should be a stopper.